### ISSUE TITLE
Add validation of cyclic inheritance for declared types

### ISSUE THREAD

**Issue by pluralia** [langium/langium#839]
Currently, no validation of this code, where type `X` extends `Y`, `Y` extends `Z` and `Z` extends `X`.
```
interface X extends Y {}
interface Y extends Z {}
interface Z extends X {}

**Comment by snarkipus** [comment]
I implemented something similar to this when I was attempting to learn how to use Langium.
<details>
  <summary>Helper Functions</summary>

```typescript
export function classHierarchy(c: SJClass): Set<Reference<SJClass>> {
    let visited = new Set<Reference<SJClass>>();
    let current = c.superClass;
    while (current != undefined && !visited.has(current)) {
        visited.add(current);
        current = current.ref?.superClass;
    }
    return visited;
}

export function classHierarchyMethods(c: SJClass): Map<string, SJMethod> {
        let returnMap = new Map<string, SJMethod>();

        [...classHierarchy(c)]
            .reverse()
            .flatMap(r => r.ref?.members as SJMember[])
            .forEach(m => returnMap.set(m.name, m as SJMethod));

        return returnMap;
}

export function classHierarchyMembers(c: SJClass): Array<SJMember> {
    return [...classHierarchy(c)]
                .map(r => r.ref)
                .map(m => m?.members as SJMember[])
                .flat();
}
```
</details>

<details>
  <summary>Validator</summary>

```typescript
    checkClassHierarchy(c: SJClass, accept: ValidationAcceptor): void {
        const hierarchy = classHierarchy(c);
        hierarchy.forEach(ref => {
            if (ref.ref?.name === c.name) {
                accept(
                    'error',
                    "cycle in hierarchy of class '" + c.name + "'",
                    {
                        node: c,
                        property: 'name',
                        code: IssueCodes.ClassHierarchy
                    }
                );
            }
        })
    }
```
</details>

<details>
  <summary>Test Code</summary>

```typescript
describe('Small Java Validator: Valid Hierarchy', () => {
    const text=`
    class A extends C {}
    class C extends B {}
    class B {}
    `;

    let validationResult: ValidationResult<SJProgram>;

    beforeAll(async () => {
        validationResult = await validate(text);
    });

    it('Should detect no cycle hierarchies', () => {
        expectNoIssues(validationResult);
    });
});

describe('Small Java Validator: Hierarchy Cycles', () => {
    const text=`
    class A extends C {}
    class C extends B {}
    class B extends A {}
    `;

    let validationResult: ValidationResult<SJProgram>;

    beforeAll(async () => {
        validationResult = await validate(text);
    });

    it('Should detect cycle hierarchies', () => {
        const rule = validationResult.document.parseResult.value.classes;
        for (let i = 0; i < 3; i++) {
            expectError(validationResult, "cycle in hierarchy of class '" + rule[i].name + "'",
            {
                node: rule[i],
                property: {
                    name: 'name'
                }
            });
        }
    });
});
```
</details>

Not sure how helpful this is since it was married to my own nascent type computation implementation, but it worked for what I was attempting to do. Hopefully this saves you some time. :)

**Comment by pluralia** [comment]
@snarkipus, thank you! Could you make a PR to close this issue?

**Comment by snarkipus** [comment]
@pluralia Sure, I'd be happy to - it will just take me a few days. I'm not sure that @msujew had a PR from me on his Christmas list. :)

**Comment by snarkipus** [comment]
Alright, so I took a look at this ... turns out it's not quite as straightforward as I hoped. 

<details>
  <summary>Grammar</summary>

```typescript
Interface:
    'interface' name=ID
    ('extends' superTypes+=[AbstractType] (',' superTypes+=[AbstractType])*)?
    SchemaType;
```
</details>

<details>
  <summary>AST</summary>

```typescript
export type AbstractType = Action | Interface | ParserRule | Type;
...
export interface Interface extends AstNode {
    readonly $container: Alternatives | Assignment | AtomType | CharacterRange | CrossReference | Grammar | Group | NegatedToken | ParserRule | TerminalAlternatives | TerminalGroup | TerminalRule | UnorderedGroup | UntilToken;
    readonly $type: 'Interface';
    attributes: Array<TypeAttribute>
    name: string
    superTypes: Array<Reference<AbstractType>>
}
```
</details>

The Langium syntax supports interfaces that can extend a seemingly arbitrary number of `AbstractType`  references (multiple-inheritance?). While I did get past the fascinating challenge of `filter`'s lack of type inference, I'm left scratching my head a bit about how best to walk the graph of supertypes. My naive attempt got messy in a hurry which usually means I've missed the obvious, idiomatic solution.

Any pointers or suggestions?

**Comment by pluralia** [comment]
@snarkipus, thank for the efforts :)

> how best to walk the graph of supertypes

One of options is [here](https://github.com/langium/langium/blob/main/packages/langium/src/grammar/validation/validation-resources-collector.ts#L71-L116) -- the function collects aliases for each type to close [this issue](https://github.com/langium/langium/issues/823). This is BFS and you need DFS, but the idea can be used:

1. Use `ValidationResources.typeToValidationInfo` to have all types
2. Filter it -- some types have inferred and declared versions: if the type has a declared version, you don't need inferred one; create a map `nameToType` from name to the real type representation to access the type faster [here](https://github.com/langium/langium/blob/main/packages/langium/src/grammar/validation/validation-resources-collector.ts#L72-L75)
4. DFS ([BFS is here](https://github.com/langium/langium/blob/main/packages/langium/src/grammar/validation/validation-resources-collector.ts#L88-L113)):
- initially you have some node of your graph -- real type representation (`queue` in the example contains `TypeOption`s)
- get edges by accessing `realSuperTypes` property of this node
- now you have names of children and can get their type representation using `nameToType`

**Comment by snarkipus** [comment]
Thanks @pluralia - I appreciate you taking the time to mentor me. Hopefully, I get partial credit for asking the right question. :)

I should definitely have enough to get started now. I'm sure I'll have more questions as I work through it.

**Comment by spoenemann** [comment]
Fixed with #1130
